File: android-app/src/analysis.ts
/**
 * analysis.ts — derive telemetry channels from a lap's raw GPS track.
 *
 * The ESP saves every 5th GPS point (~2 Hz) as [lat, lon, lap_ms].
 * We can derive:
 *   - distance along the track (cumulative meters)
 *   - speed (km/h, from position/time derivative)
 *   - longitudinal G (change of speed over time)
 *   - lateral G (change of heading over time × speed)
 *   - delta-to-reference (time difference at the same track distance)
 *
 * All functions are pure — they take a Lap and return arrays sampled at
 * the same indices as lap.track_points.
 */

import { Lap } from './types';

const DEG2RAD = Math.PI / 180;
const EARTH_R = 6371000;  // meters

/** Haversine distance in meters between two GPS points. */
export function haversineM(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const dLat = (lat2 - lat1) * DEG2RAD;
  const dLon = (lon2 - lon1) * DEG2RAD;
  const a = Math.sin(dLat / 2) ** 2
          + Math.cos(lat1 * DEG2RAD) * Math.cos(lat2 * DEG2RAD) * Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_R * Math.asin(Math.sqrt(a));
}

/** Heading (true north = 0) from p1 to p2 in degrees [0, 360). */
function bearingDeg(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const φ1 = lat1 * DEG2RAD, φ2 = lat2 * DEG2RAD;
  const Δλ = (lon2 - lon1) * DEG2RAD;
  const y = Math.sin(Δλ) * Math.cos(φ2);
  const x = Math.cos(φ1) * Math.sin(φ2) - Math.sin(φ1) * Math.cos(φ2) * Math.cos(Δλ);
  return (Math.atan2(y, x) / DEG2RAD + 360) % 360;
}

export interface LapChannels {
  /** ms since lap start, length N */
  t_ms:  number[];
  /** cumulative distance in meters, length N */
  dist_m: number[];
  /** speed in km/h, length N */
  speed_kmh: number[];
  /** longitudinal G (+ = accel, - = braking), length N */
  g_long: number[];
  /** lateral G (+ = right turn, - = left), length N */
  g_lat: number[];
  /** heading in degrees, length N */
  heading: number[];
}

/** Compute all derived channels for one lap. Returns empty arrays if <2 points. */
export function deriveChannels(lap: Lap): LapChannels {
  const pts = lap.track_points ?? [];
  const N = pts.length;
  const out: LapChannels = {
    t_ms: [], dist_m: [], speed_kmh: [], g_long: [], g_lat: [], heading: [],
  };
  if (N < 2) return out;

  // Time + cumulative distance
  let cum = 0;
  out.t_ms.push(pts[0][2]);
  out.dist_m.push(0);
  for (let i = 1; i < N; i++) {
    const [lat0, lon0] = pts[i - 1];
    const [lat1, lon1, ms1] = pts[i];
    cum += haversineM(lat0, lon0, lat1, lon1);
    out.t_ms.push(ms1);
    out.dist_m.push(cum);
  }

  // Speed (km/h) at each point: central differences, except endpoints.
  out.speed_kmh.push(0);
  for (let i = 1; i < N - 1; i++) {
    const dd = out.dist_m[i + 1] - out.dist_m[i - 1];
    const dt_s = (out.t_ms[i + 1] - out.t_ms[i - 1]) / 1000;
    out.speed_kmh.push(dt_s > 0 ? (dd / dt_s) * 3.6 : 0);
  }
  out.speed_kmh.push(out.speed_kmh[N - 2] ?? 0);

  // Heading: from each segment, clamp to point index by using the bearing
  // from previous to current point.
  out.heading.push(0);
  for (let i = 1; i < N; i++) {
    out.heading.push(bearingDeg(pts[i - 1][0], pts[i - 1][1], pts[i][0], pts[i][1]));
  }

  // Longitudinal G: dv/dt, /9.81
  out.g_long.push(0);
  for (let i = 1; i < N - 1; i++) {
    const dv = (out.speed_kmh[i + 1] - out.speed_kmh[i - 1]) / 3.6;  // m/s
    const dt = (out.t_ms[i + 1] - out.t_ms[i - 1]) / 1000;
    out.g_long.push(dt > 0 ? dv / dt / 9.81 : 0);
  }
  out.g_long.push(0);

  // Lateral G: v² × curvature. Curvature ≈ Δheading / distance.
  out.g_lat.push(0);
  for (let i = 1; i < N - 1; i++) {
    let dHead = out.heading[i + 1] - out.heading[i - 1];
    // Wrap to [-180, 180]
    while (dHead > 180)  dHead -= 360;
    while (dHead < -180) dHead += 360;
    const dHeadRad = dHead * DEG2RAD;
    const ds = out.dist_m[i + 1] - out.dist_m[i - 1];
    const v = out.speed_kmh[i] / 3.6;  // m/s
    const kappa = ds > 0 ? dHeadRad / ds : 0;
    out.g_lat.push((v * v * kappa) / 9.81);
  }
  out.g_lat.push(0);

  // Light smoothing pass on G-channels (3-tap moving average) to reduce noise.
  out.g_long = smooth3(out.g_long);
  out.g_lat  = smooth3(out.g_lat);
  out.speed_kmh = smooth3(out.speed_kmh);

  return out;
}

function smooth3(arr: number[]): number[] {
  const n = arr.length;
  if (n < 3) return arr.slice();
  const out = new Array(n);
  out[0] = arr[0];
  out[n - 1] = arr[n - 1];
  for (let i = 1; i < n - 1; i++) {
    out[i] = (arr[i - 1] + arr[i] + arr[i + 1]) / 3;
  }
  return out;
}

/**
 * Delta-to-reference. For each distance d along the TARGET lap, find the time
 * it took the REFERENCE lap to reach that same distance. Delta = target_time
 * - reference_time (positive means target is slower).
 *
 * Returns delta in ms, one value per target-lap sample.
 */
export function deltaToReference(target: LapChannels, reference: LapChannels): number[] {
  const out: number[] = [];
  const refN = reference.dist_m.length;
  if (refN < 2) return target.t_ms.map(() => 0);

  let j = 0;  // running index into reference
  for (let i = 0; i < target.t_ms.length; i++) {
    const d = target.dist_m[i];
    // advance j while reference distance is still < d
    while (j < refN - 1 && reference.dist_m[j + 1] < d) j++;
    // interpolate reference time at distance d
    const d0 = reference.dist_m[j];
    const d1 = reference.dist_m[Math.min(j + 1, refN - 1)];
    const t0 = reference.t_ms[j];
    const t1 = reference.t_ms[Math.min(j + 1, refN - 1)];
    let ref_t: number;
    if (d1 <= d0) {
      ref_t = t1;
    } else if (d >= d1) {
      ref_t = t1;  // target went further than reference — cap
    } else {
      const frac = (d - d0) / (d1 - d0);
      ref_t = t0 + frac * (t1 - t0);
    }
    out.push(target.t_ms[i] - ref_t);
  }
  return out;
}

/**
 * Min/max across a set of series — for Y-axis auto-scale.
 * Pads the range by 5 % to avoid lines touching the chart edge.
 */
export function yRange(series: number[][]): [number, number] {
  let min = Infinity, max = -Infinity;
  for (const s of series) {
    for (const v of s) {
      if (!isFinite(v)) continue;
      if (v < min) min = v;
      if (v > max) max = v;
    }
  }
  if (!isFinite(min) || !isFinite(max)) return [0, 1];
  if (min === max) { min -= 0.5; max += 0.5; }
  const pad = (max - min) * 0.05;
  return [min - pad, max + pad];
}

/** Nicely format a duration (ms) as "1:23.45" or "23.45s" */
export function fmtLapTime(ms: number): string {
  if (!isFinite(ms) || ms <= 0) return '—';
  const m = Math.floor(ms / 60000);
  const s = (ms % 60000) / 1000;
  if (m > 0) return `${m}:${s.toFixed(2).padStart(5, '0')}`;
  return `${s.toFixed(2)}s`;
}

/** Format delta ms as "+0.34 s" or "-1.02 s" */
export function fmtDelta(ms: number): string {
  if (!isFinite(ms) || ms === 0) return '±0.00 s';
  const sign = ms > 0 ? '+' : '';
  return `${sign}${(ms / 1000).toFixed(2)} s`;
}

File: tools/brl_studio/core/analysis.py
"""
Math channels derived from raw GPS track points.

Mirrors the algorithms from android-app/src/lib/analysis.ts so the desktop
Studio shows the same numbers as the phone for the same lap. When the
firmware starts persisting OBD/CAN/analog channels in the session JSON,
those flow through directly without going through this module.

Distance/speed use the equirectangular approximation — accurate to <0.5%
at racing-circuit scales (<10 km) and orders of magnitude faster than
haversine; fine for charts.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from .session import Lap, TrackPoint

EARTH_R_M = 6_371_000.0


def _flat_distance_m(p1: TrackPoint, p2: TrackPoint) -> float:
    lat_avg = math.radians((p1.lat + p2.lat) * 0.5)
    dlat = math.radians(p2.lat - p1.lat)
    dlon = math.radians(p2.lon - p1.lon) * math.cos(lat_avg)
    return EARTH_R_M * math.hypot(dlat, dlon)


@dataclass
class LapChannels:
    """Per-point arrays, all len == len(lap.points)."""

    distance_m: list[float]      # cumulative from lap start
    speed_kmh: list[float]
    g_long: list[float]          # +accel / -brake (g)
    g_lat: list[float]           # cornering g
    heading_deg: list[float]


def _smooth3(a: list[float]) -> list[float]:
    """3-tap centered moving average. Endpoints are passed through.
    Knocks down isolated spikes by 3× without losing real features."""
    n = len(a)
    if n < 3:
        return list(a)
    out = [0.0] * n
    out[0] = a[0]
    out[-1] = a[-1]
    for i in range(1, n - 1):
        out[i] = (a[i - 1] + a[i] + a[i + 1]) / 3.0
    return out


def derive_channels(lap: Lap) -> LapChannels:
    """Speed / G / heading from a lap's track points.

    Algorithm matches the Android app's analysis.ts -- proven to work
    cleanly with the firmware's ~10 Hz GPS stream. The key tricks:
        * Dedup consecutive samples with identical (lat, lon). Firmware
          ticks faster than GPS gives new fixes, so the same position
          is often stored 2-3× back-to-back with rising lap_ms. Naively
          differentiating those produces a 0 / V / 0 / V sawtooth that
          dominates the chart -- exactly the aliasing the user reported.
          Keeping ONLY samples where the position has actually moved
          collapses the duplicates back into a clean d/dt that reflects
          the real GPS update cadence (5-10 Hz adaptive).
        * Cumulative distance via flat-Earth approximation per step.
        * Speed via CENTRAL differences -- (i+1 vs i-1) instead of
          backward (i vs i-1). Halves the per-sample noise sensitivity.
        * Hard cap on physically unreasonable speed values to defang
          single-sample GPS glitches before they propagate into G calc.
        * Final 3-tap moving-average smoothing on speed + both Gs to
          remove residual jitter without hiding real braking / cornering.
    """
    SPEED_MAX_KMH = 400.0
    G_MAX = 4.0
    # ~1 m at typical lats. Anything below this between two samples is
    # GPS jitter, not motion -- treat both samples as the same position.
    DEDUP_LATLON_EPS = 1e-5

    n_raw = len(lap.points)
    if n_raw == 0:
        return LapChannels([], [], [], [], [])

    # Dedup pass: keep first sample, then only samples that moved more
    # than DEDUP_LATLON_EPS from the LAST KEPT sample. lap_ms is taken
    # from each kept sample, so dt naturally widens to span the silent
    # period between real GPS fixes.
    pts = [lap.points[0]]
    for p in lap.points[1:]:
        last = pts[-1]
        if (abs(p.lat - last.lat) > DEDUP_LATLON_EPS
                or abs(p.lon - last.lon) > DEDUP_LATLON_EPS):
            pts.append(p)
    # Mutate lap.points so downstream callers (map plotting, cursor sync
    # via index lookup) stay aligned with the channels arrays. Without
    # this the analyse view's `lap.points[idx]` would walk into the
    # original duplicate-laden list while idx is in the dedupped space,
    # and the map cursor would land at the wrong physical spot.
    lap.points = pts
    n = len(pts)
    t_ms = [float(p.lap_ms) for p in pts]

    # Cumulative distance
    distance = [0.0] * n
    for i in range(1, n):
        distance[i] = distance[i - 1] + _flat_distance_m(pts[i - 1], pts[i])

    # Heading per point: bearing of incoming segment
    heading = [0.0] * n
    for i in range(1, n):
        dy = math.radians(pts[i].lat - pts[i - 1].lat)
        dx = math.radians(pts[i].lon - pts[i - 1].lon) * math.cos(
            math.radians(pts[i].lat))
        heading[i] = math.degrees(math.atan2(dx, dy)) % 360.0
    if n >= 2:
        heading[0] = heading[1]

    # Speed via central differences -- much less noisy than backward diff.
    speed = [0.0] * n
    for i in range(1, n - 1):
        dd = distance[i + 1] - distance[i - 1]
        dt_s = (t_ms[i + 1] - t_ms[i - 1]) / 1000.0
        if dt_s > 0:
            v = (dd / dt_s) * 3.6
            if v < 0 or v > SPEED_MAX_KMH:
                v = speed[i - 1]
            speed[i] = v
    if n >= 2:
        speed[0] = speed[1]
        speed[-1] = speed[-2]

    # Longitudinal g: dv/dt, central differences too.
    g_long = [0.0] * n
    for i in range(1, n - 1):
        dv_ms = (speed[i + 1] - speed[i - 1]) / 3.6
        dt_s = (t_ms[i + 1] - t_ms[i - 1]) / 1000.0
        if dt_s > 0:
            g = dv_ms / dt_s / 9.80665
            if g < -G_MAX or g > G_MAX:
                g = g_long[i - 1]
            g_long[i] = g

    # Lateral g: v² × curvature, where curvature ≈ Δheading / distance.
    g_lat = [0.0] * n
    for i in range(1, n - 1):
        dh = heading[i + 1] - heading[i - 1]
        while dh > 180:
            dh -= 360
        while dh < -180:
            dh += 360
        ds = distance[i + 1] - distance[i - 1]
        v_ms = speed[i] / 3.6
        if ds > 0:
            kappa = math.radians(dh) / ds
            g = (v_ms * v_ms * kappa) / 9.80665
            if g < -G_MAX or g > G_MAX:
                g = g_lat[i - 1]
            g_lat[i] = g

    # Final smoothing pass — same 3-tap MA the Android app uses.
    speed = _smooth3(speed)
    g_long = _smooth3(g_long)
    g_lat = _smooth3(g_lat)

    return LapChannels(
        distance_m=distance,
        speed_kmh=speed,
        g_long=g_long,
        g_lat=g_lat,
        heading_deg=heading,
    )


def delta_vs_reference(lap: Lap, ref: Lap) -> list[float]:
    """
    Per-point time delta in seconds: lap_time_at_position - ref_time_at_position.
    Positive = slower than reference. Uses nearest-point matching with a
    forward-monotonic cursor (same idea as live_delta.cpp on the device).
    """
    if not lap.points or not ref.points:
        return []
    cursor = 0
    out: list[float] = []
    ref_n = len(ref.points)
    for p in lap.points:
        # Forward search around cursor (don't allow long backwards jumps)
        best_i = cursor
        best_d = float("inf")
        end = min(ref_n - 1, cursor + 60)
        for i in range(cursor, end + 1):
            r = ref.points[i]
            dlat = p.lat - r.lat
            dlon = (p.lon - r.lon) * math.cos(math.radians(p.lat))
            d2 = dlat * dlat + dlon * dlon
            if d2 < best_d:
                best_d = d2
                best_i = i
        cursor = best_i
        out.append((p.lap_ms - ref.points[best_i].lap_ms) / 1000.0)
    return out
